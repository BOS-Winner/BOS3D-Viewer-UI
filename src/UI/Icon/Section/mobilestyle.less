@import "../../theme/index";


.container{
  :global(.mobile-device) &{
       @width: 68px;
       @height: 56px;
      .sectionGroup{
        display: flex;
        flex-direction: row;
        justify-content: space-around;
        align-items: center;
        svg{
          width: 32px;
          height: 32px;
        }     
      }

      .divider{
        width: 100%;
        height: 1px;
        background: rgba(255, 255, 255, 0.16);
        margin: 10px 0;
      }
      .h3{
        margin: 0;
        padding: 0;
        font-size: 16px;
        font-family: PingFangSC-Regular, PingFang SC;
        color: #FFFFFF;
      }

      
      .free-wrapper{
        display: flex;
        flex-direction: row;
        justify-content: space-around;
        align-items: center;
        @height: 60px;
        margin: 10px 0 0 0;
        .select,  .eye{
          background: #363A3D;
          border-radius: 2px;
          border: 1px solid rgba(255, 255, 255, 0.1);
          outline: none;
          width: @width;
          height: @height;
          padding: 0;
          margin: 0;
          display: flex;
          align-items: center;
          justify-content: center;
          font-size: 18px;
          font-family: PingFangSC-Regular, PingFang SC;
          font-weight: 400;
          color: #DDDDDD;
          &:active{
            opacity:0.5
          } ;
        }
        .eye{
          svg{
            width: 32px;
            height: 32px;
          }    
        }
      }

      .btnGroup{
        :global{
          .bos-btn{
            border-radius: 4px;
            border: 1px solid rgba(255, 255, 255, 0.1);
            margin-top: 12px;
            display: flex;
            justify-items: center;
            justify-content: center;
            align-items: center;
            height: 32px;
          }
          .bos-btn-default{
            background-color:#707477;
         
          }
          svg{
            width: 20px;
            height: 20px;
            margin-right: 5px;
          } 
        }
      }
  }
}