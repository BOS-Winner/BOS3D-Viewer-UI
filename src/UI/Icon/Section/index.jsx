import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import Modal from "Base/Modal";
import Card from "Base/Card";
import { Select } from "antd";
import { changeMode } from "../action";
import Icon from "../../Base/Icon";
import toastr from "../../toastr";
import style from "./style.less";
import { AntdIcon } from "../../utils/utils";
import iconstyle from "../../Toolbar/bottom.less";
import mobilestyle from "./mobilestyle.less";
import { DEFAULT_MODAL_PLACE } from "../../constant";
// import MeasureControl from '../Measure/MobileHelper/MeasureControl';

class SectionPopup extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      showModal: false,
      mode: props.mode,
      showPlane: true,
      isShowSectionBox: true,
      selectDir: '',
      touchProgress: 50,
      leftRightRotation: 0,
      upDownRoataion: 0,
    };
    this.sectionPlane = null;
    // this.touchProgress = 50;
    // this.leftRightRotation = 0;
    // this.upDownRoataion = 0;
    this.isTouchDevice = props.BIMWINNER.BOS3D.DeviceTest.isTouchDevice();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.mode !== "剖切模式" && this.props.mode === "剖切模式") {
      // 切换模式时提示用户
      if (prevProps.mode.length > 0) {
        toastr.remove();
        toastr.info(`剖切模式下将关闭${prevProps.mode}`, "", {
          target: `#${this.props.viewer.viewport}`,
        });
      }
    }
  }

  // eslint-disable-next-line react/no-deprecated
  componentWillReceiveProps(prevProps) {
    // 如果由当前的剖切模式切换到其他模式则关闭当前的剖切模式
    if (prevProps.mode !== "剖切模式" && this.props.mode === "剖切模式") {
      // 关闭模型剖切
      if (this.state.mode === "剖切模式") {
        this.props.viewer.disableSectionBox();
      }
      // 关闭自由剖切模式
      if (this.sectionPlane) {
        this.sectionPlane.exit();
        this.sectionPlane = null;
        this.setState((state) => ({
          ...state,
          mode: "",
          showModal: false,
        }));
      }
      this.props.viewer.render();
      this.setState((state) => ({
        ...state,
        mode: "",
        showModal: false,
      }));
    }
  }

  /**
   * 开启自由剖切
   */
  startFreeSection = () => {
    const BIMWINNER = this.props.BIMWINNER;
    this.sectionPlane = new BIMWINNER.BOS3D.SectionPlane({
      direction: "Forward",
      id: "SectionPlane",
      plane: "X",
      progress: 50,
      viewer: this.props.viewer,
    });
    if (this.isTouchDevice) {
      this.sectionPlane.coordinateSystem.hide(true);
    }
    this.props.viewer.render();
  };

  /**
   * 开启关闭剖切弹窗
   */
  triggerPopUp = () => {
    const mode = this.props.mode;
    if (mode === "剖切模式") {
      this.props.changeMode("");
      // 关闭模型剖切
      if (this.state.mode === "剖切模式") {
        this.props.viewer.disableSectionBox();
        this.props.viewer.render();
      }
      // 关闭自由剖切模式
      if (this.sectionPlane) {
        this.sectionPlane.exit();
        this.sectionPlane = null;
      }
      this.setState((state) => ({
        ...state,
        mode: "",
      }));
    } else {
      this.props.changeMode("剖切模式");
    }
    this.setState((state) => ({
      ...state,
      showModal: !state.showModal,
    }));
  };

  /**
   * 切换剖切模式
   * @param {string} type 剖切模式
   */
  switchSection = (type) => {
    this.state.selectDir = '';
    if (type === "自由剖切模式") {
      // this.props.changeMode("自由剖切模式");
      this.setState((state) => ({
        ...state,
        showPlane: true,
        mode: "自由剖切模式",
      }));
      // 关闭模型剖切
      this.props.viewer.disableSectionBox();
      this.props.viewer.render();
      // 开启自由剖切
      this.startFreeSection();
    } else {
      // this.props.changeMode("剖切模式");
      this.setState((state) => ({
        ...state,
        isShowSectionBox: true,
        mode: "剖切模式",
      }));
      // 关闭自由剖切模式
      if (this.sectionPlane) {
        this.sectionPlane.exit();
        this.sectionPlane = null;
      }
      // 开启模型剖切额模式
      this.props.viewer.enableSectionBox();
      this.props.viewer.setSectionBoxMode("normal");
      this.props.viewer.render();
    }
  };

  /**
   * 选择轴向
   * @param {string}} value 选择轴向
   */
  handleSelect = (value) => {
    const type = value;
    this.setState({
      selectDir: type
    });
    if (type[0] === "-") {
      this.sectionPlane.setDirection("Reverse");
      this.sectionPlane.setPlane(type[1]);
    } else {
      this.sectionPlane.setDirection("Forward");
      this.sectionPlane.setPlane(type);
    }
    if (this.isTouchDevice) {
      this.sectionPlane.setProgress(this.state.touchProgress);
    }
    this.setState({
      touchProgress: 50,
      leftRightRotation: 0,
      upDownRoataion: 0
    });
    this.props.viewer.render();
  };

  /**
   * 是否显示
   * @param {object} e 事件
   */
  switchShowPlane = (e) => {
    e.preventDefault();
    e.stopPropagation();
    this.setState((state) => {
      if (state.showPlane) {
        this.sectionPlane.hidePlane();
      } else {
        this.sectionPlane.showPlane();
      }
      return {
        showPlane: !state.showPlane,
      };
    });
  };

  toggleSectionBox() {
    const visible = this.state.isShowSectionBox;
    if (visible) {
      this.props.viewer.hideSectionBox();
    } else {
      this.props.viewer.showSectionBox();
    }
    this.setState({
      isShowSectionBox: !visible,
    });
  }

  resetSectionBox() {
    this.props.viewer.resetSectionBox();
    this.setState({
      isShowSectionBox: true,
    });
  }

  onTouchMove = (clientX, clientY) => {
    this.props.viewer.getViewerImpl().controlManager.getToolByName('pickByMeasure').touchPointMove({
      clientX,
      clientY
    });
  }

  setProgress = (e) => {
    e.stopPropagation();
    this.touchProgress = Number(e.target.value);
    this.setState({
      touchProgress: Number(e.target.value)
    });
    this.sectionPlane.setProgress(this.touchProgress);
    this.props.viewer.render();
  }

  // 左右旋转
  customRotation = (a, b) => {
    this.leftRightRotation = a;
    this.upDownRoataion = b;
    this.setState({
      leftRightRotation: a,
      upDownRoataion: b
    });
    this.sectionPlane.setRotateAngle(a, b);
    this.props.viewer.render();
  }

  render() {
    const { showModal, mode } = this.state;
    const { isMobile } = this.props;
    if (isMobile) {
      return this.renderMobileJsx();
    }
    const sectionMode = [
      {
        icon: "iconziyoupouqie-01",
        title: "自由剖切",
        func: () => this.switchSection("自由剖切模式"),
        active: mode === "自由剖切模式",
      },
      {
        icon: "iconmoxingpouqie-01",
        title: "模型剖切",
        func: () => this.switchSection("剖切模式"),
        active: mode === "剖切模式",
      },
    ];

    const freeItem = [
      {
        value: "X",
        text: "X轴",
      },
      {
        value: "Y",
        text: "Y轴",
      },
      {
        value: "Z",
        text: "Z轴",
      },
      {
        value: "-X",
        text: "-X轴",
      },
      {
        value: "-Y",
        text: "-Y轴",
      },
      {
        value: "-Z",
        text: "-Z轴",
      },
    ];
    return (
      <div id="customSection">
        <Modal
          visible={showModal}
          onCancel={this.triggerPopUp}
          title="剖切"
          width="350px"
          height="220px"
          top={DEFAULT_MODAL_PLACE.section.top}
          right={DEFAULT_MODAL_PLACE.section.right}
          left={DEFAULT_MODAL_PLACE.section.left}
          minWidth={350}
          minHeight={220}
          viewportDiv={this.props.viewer.viewportDiv}
        >
          <div className={style.sectionGroup}>
            {sectionMode.map((item) => (
              <Card
                key={item.title}
                icon={item.icon}
                title={item.title}
                func={item.func}
                active={item.active}
              />
            ))}
          </div>
          <div className={style.divider} />
          {mode === "自由剖切模式" ? (
            <div className={style.freeWrapper}>
              <Select
                className={style.select}
                onChange={this.handleSelect}
                dropdownClassName={style.dropDown}
                defaultValue="X"
              >
                {freeItem.map((item) => (
                  <Select.Option
                    key={item.value}
                    value={item.value}
                    className={style.option}
                  >
                    {item.text}
                  </Select.Option>
                ))}
              </Select>
              <div
                title={
                  this.state.showPlane ? "点击隐藏剖切面" : "点击显示剖切面"
                }
                className={style.eye}
                role="presentation"
                onClick={this.switchShowPlane}
              >
                <AntdIcon
                  type={
                    this.state.showPlane
                      ? "iconcaozuolishikejian-01"
                      : "iconcaozuolishibukejian-01"
                  }
                />
              </div>
            </div>
          ) : null}
          {mode === "剖切模式" ? (
            <div className={style.btnGroup}>
              <div role="presentation" onClick={() => this.toggleSectionBox()}>
                {this.state.isShowSectionBox ? "隐藏剖切框" : "显示剖切框"}
              </div>
              <div role="presentation" onClick={() => this.resetSectionBox()}>
                重置剖切框
              </div>
            </div>
          ) : null}
        </Modal>

        <div
          title="剖切"
          role="button"
          tabIndex={0}
          onClick={() => {
            this.triggerPopUp();
          }}
        >
          <Icon
            icon={
              <AntdIcon type="iconmoxingpouqie-01" className={iconstyle.icon} />
            }
            selected={this.state.showModal}
            title="剖切"
          />
        </div>
      </div>
    );
  }

  renderMobileJsx() {
    const { showModal, mode, selectDir } = this.state;
    // const { isMobile } = this.props;
    const sectionMode = [
      {
        icon: "iconziyoupouqie-01",
        title: "自由剖切",
        func: () => this.switchSection("自由剖切模式"),
        active: mode === "自由剖切模式",
      },
      {
        icon: "iconmoxingpouqie-01",
        title: "模型剖切",
        func: () => this.switchSection("剖切模式"),
        active: mode === "剖切模式",
      },
    ];

    const freeItem = [
      {
        value: "X",
        text: "X轴",
      },
      {
        value: "Y",
        text: "Y轴",
      },
      {
        value: "Z",
        text: "Z轴",
      },
      {
        value: "-X",
        text: "-X轴",
      },
      {
        value: "-Y",
        text: "-Y轴",
      },
      {
        value: "-Z",
        text: "-Z轴",
      },
    ];

    const customStyle = {
      width: '68px',
      height: '56px'
    };
    const selectElementStyle = {
      paddingLeft: selectDir.length > 1 ? 12 : 20
    };
    return (
      <div id="customSection">
        {/* {mode === "自由剖切模式" && (
          <MeasureControl
            onTouchMove={()=>{}}
            ok={() => {}}
            cancel={() => {}}
            viewer={this.props.viewer}
          />
        )} */}
        <Modal
          visible={showModal}
          onCancel={this.triggerPopUp}
          title="剖切"
          width="176px"
          minWidth={176}
          minHeight={220}
          viewportDiv={this.props.viewer.viewportDiv}
          theme="mobile-theme-one"
        >
          <section className={mobilestyle.container}>
            <div className={mobilestyle.sectionGroup}>
              {sectionMode.map((item) => (
                <Card
                  notHover
                  key={item.title}
                  icon={item.icon}
                  func={item.func}
                  active={item.active}
                  customStyle={customStyle}
                />
              ))}
            </div>
            <div className={mobilestyle.divider} />
            {mode === "自由剖切模式" ? (
              <>
                <h3 className={mobilestyle.h3}>自由剖切</h3>
                <div className={mobilestyle.freeWrapper}>
                  <select
                    style={selectElementStyle}
                    className={mobilestyle.select}
                    onChange={(e) => {
                      this.handleSelect(e.target.value);
                    }}
                    defaultValue="X"
                  >
                    {freeItem.map((item) => (
                      <option
                        key={item.value}
                        value={item.value}
                        className={mobilestyle.option}
                      >
                        {item.text}
                      </option>
                    ))}
                  </select>
                  <div
                    title={
                      this.state.showPlane ? "点击隐藏剖切面" : "点击显示剖切面"
                    }
                    className={mobilestyle.eye}
                    role="presentation"
                    onClick={this.switchShowPlane}
                  >
                    <AntdIcon
                      type={
                        this.state.showPlane
                          ? "iconcaozuolishikejian-01"
                          : "iconcaozuolishibukejian-01"
                      }
                    />
                  </div>
                </div>
                <span className={style.customSubTitle} style={{ marginTop: '10px' }}>轴向距离</span>
                <div className={style.customInput}>
                  <span>0</span>
                  <div className={style.range}>
                    <input
                      type="range"
                      min={0}
                      max={100}
                      step={1}
                      value={this.state.touchProgress}
                      onChange={e => { this.setProgress(e) }}
                    />
                  </div>
                  <span>100%</span>
                </div>
                <span className={style.customSubTitle}>左右旋转</span>
                <div className={style.customInput}>
                  <span>0</span>
                  <div className={style.range}>
                    <input
                      type="range"
                      min={0}
                      max={360}
                      step={1}
                      value={this.state.leftRightRotation}
                      onChange={e => { this.customRotation(Number(e.target.value), this.state.upDownRoataion) }}
                    />
                  </div>
                  <span>360°</span>
                </div>
                <span className={style.customSubTitle}>上下旋转</span>
                <div className={style.customInput}>
                  <span>0</span>
                  <div className={style.range}>
                    <input
                      type="range"
                      min={0}
                      max={360}
                      step={1}
                      value={this.state.upDownRoataion}
                      onChange={e => { this.customRotation(this.state.leftRightRotation, Number(e.target.value)) }}
                    />
                  </div>
                  <span>360°</span>
                </div>
              </>
            ) : null}
            {mode === "剖切模式" ? (
              <>
                <h3 className={mobilestyle.h3}>模型剖切</h3>
                <div className={mobilestyle.btnGroup}>
                  <div role="presentation" className="bos-btn bos-btn-primary" onClick={() => this.toggleSectionBox()}>
                    <AntdIcon type="iconicon_yincangpouqiekuang-01" />
                    {this.state.isShowSectionBox ? "隐藏剖切框" : "显示剖切框"}
                  </div>
                  <div role="presentation" className="bos-btn bos-btn-default" onClick={() => this.resetSectionBox()}>
                    <AntdIcon type="iconicon_zhongzhipouqiekuang-" />
                    重置剖切框
                  </div>
                </div>
              </>
            ) : null}
          </section>

        </Modal>

        <div
          title="剖切"
          role="button"
          tabIndex={0}
          onClick={() => {
            this.triggerPopUp();
          }}
        >
          <Icon
            icon={
              <AntdIcon type="iconmoxingpouqie-01" className={iconstyle.icon} />
            }
            selected={this.state.showModal}
            title="剖切"
          />
        </div>
      </div>
    );
  }
}

SectionPopup.propTypes = {
  changeMode: PropTypes.func.isRequired,
  mode: PropTypes.string.isRequired,
  viewer: PropTypes.object.isRequired,
  BIMWINNER: PropTypes.object.isRequired,
  isMobile: PropTypes.bool.isRequired
};

const mapStateToProps = (state) => ({
  mode: state.button.mode,
  viewer: state.system.viewer3D,
  BIMWINNER: state.system.BIMWINNER,
  isMobile: state.system.isMobile,
});
const mapDispatchToProps = (dispatch) => ({
  changeMode: (mode) => {
    dispatch(changeMode(mode));
  },
});
const WrappedContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(SectionPopup);

export default WrappedContainer;
