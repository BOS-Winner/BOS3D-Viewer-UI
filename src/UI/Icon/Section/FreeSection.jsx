import React from "react";
import ReactDOM from "react-dom";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import Icon from "Base/Icon";
import Modal from "Base/Modal";
import { changeMode } from "../action";
import toastr from "../../toastr";
import style from "./style.less";
import fuckIE from "../../theme/fuckIE.less";
import showPng from "../img/white/show@2x.png";
import hidePng from "../img/white/hide@2x.png";
import FreeSectionPng from "../img/white/freeSection.png";

class FreeSection extends React.Component {
  constructor(props) {
    super(props);
    this.sectionPlane = {};
    this.minWidth = props.viewer.viewportDiv.clientWidth < 760 ? 120 : 170;
    this.isTouchDevice = props.BIMWINNER.BOS3D.DeviceTest.isTouchDevice();
    this.touchProgress = 50;
    this.state = {
      dialogVisible: false,
      showPlane: true,
    };
  }

  componentDidUpdate(prevProps) {
    // 开启自由剖切
    if (prevProps.mode !== "自由剖切模式" && this.props.mode === "自由剖切模式") {
      if (prevProps.mode !== '') {
        toastr.info(`自由剖切模式下将关闭${prevProps.mode}`, "", {
          target: `#${this.props.viewer.viewport}`
        });
      }
      this.startFreeSection();
      this.props.onEnter();
    }
    // 关闭自由剖切
    if (prevProps.mode === "自由剖切模式" && this.props.mode !== "自由剖切模式") {
      this.sectionPlane.exit();
    }
  }

  changeDialog(visible) {
    if (!visible) {
      this.props.changeMode('');
    }
    this.setState({
      dialogVisible: visible
    });
  }

  onClick() {
    const selected = (this.props.mode === "自由剖切模式");
    this.setState(() => {
      this.props.changeMode(selected ? "" : "自由剖切模式");
      return {
        dialogVisible: !selected,
      };
    });
  }

  startFreeSection() {
    const BIMWINNER = this.props.BIMWINNER;
    this.sectionPlane = new BIMWINNER.BOS3D.SectionPlane({
      direction: "Forward",
      id: "SectionPlane",
      plane: "X",
      progress: 50,
      viewer: this.props.viewer
    });
    if (this.isTouchDevice) {
      this.sectionPlane.coordinateSystem.hide(true);
    }
    this.props.viewer.render();
  }

  setSectionPlane(e) {
    e.stopPropagation();
    const type = e.target.value;
    if (type[0] === '-') {
      this.sectionPlane.setDirection("Reverse");
      this.sectionPlane.setPlane(type[1]);
    } else {
      this.sectionPlane.setDirection("Forward");
      this.sectionPlane.setPlane(type);
    }
    if (this.isTouchDevice) {
      this.sectionPlane.setProgress(this.touchProgress);
    }
    this.props.viewer.render();
  }

  switchShowPlane(e) {
    e.preventDefault();
    e.stopPropagation();
    this.setState(state => {
      if (state.showPlane) {
        this.sectionPlane.hidePlane();
      } else {
        this.sectionPlane.showPlane();
      }
      return {
        showPlane: !state.showPlane,
      };
    });
  }

  setProgress(e) {
    e.stopPropagation();
    this.touchProgress = Number(e.target.value);
    this.sectionPlane.setProgress(this.touchProgress);
    this.props.viewer.render();
  }

  render() {
    const selected = (this.props.mode === "自由剖切模式");
    const imgTitle = this.state.showPlane ? "点击隐藏剖切面" : "点击显示剖切面";

    const selectDirection = (
      <select
        title="改变轴向"
        defaultValue="X"
        onChange={e => {
          this.setSectionPlane(e);
        }}
      >
        <option value="X">X</option>
        <option value="Y">Y</option>
        <option value="Z">Z</option>
        <option value="-X">-X</option>
        <option value="-Y">-Y</option>
        <option value="-Z">-Z</option>
      </select>
    );
    const showPlane = (
      <img
        className={style.eye}
        src={this.state.showPlane ? showPng : hidePng}
        alt={imgTitle}
        title={imgTitle}
        onClick={e => {
          this.switchShowPlane(e);
        }}
      />
    );

    let content = <></>;
    if (this.isTouchDevice) {
      content = ReactDOM.createPortal((
        <div
          role="presentation"
          className={style.touchBar}
          style={{
            display: this.state.dialogVisible && selected ? "flex" : "none"
          }}
          onClick={e => {
            e.preventDefault();
            e.stopPropagation();
          }}
        >
          {selectDirection}
          {showPlane}
          <span>0</span>
          <div className={style.range}>
            <input
              type="range"
              min={0}
              max={100}
              step={1}
              defaultValue={this.touchProgress}
              onChange={e => { this.setProgress(e) }}
            />
          </div>
          <span>100</span>
          <button
            type="button"
            onClick={e => {
              e.preventDefault();
              e.stopPropagation();
              this.changeDialog(false);
            }}
          >
            退出
          </button>
        </div>
      ), this.props.viewer.viewportDiv);
    } else {
      content = (
        <Modal
          visible={this.state.dialogVisible && selected}
          onCancel={() => {
            this.changeDialog(false);
          }}
          title="自由剖切选项"
          width={`${this.minWidth}px`}
          height="110px"
          destroyOnClose
          minWidth={this.minWidth}
          minHeight={100}
          viewportDiv={this.props.viewer.viewportDiv}
        >
          <div className={style.freeSection}>
            {selectDirection}
            {showPlane}
          </div>
        </Modal>
      );
    }
    return (
      <div
        className={selected ? fuckIE.select : ''}
        title="自由剖切"
        role="button"
        tabIndex={0}
        onClick={() => {
          this.onClick();
        }}
      >
        <Icon img={FreeSectionPng} />
        {content}
      </div>
    );
  }
}

FreeSection.propTypes = {
  changeMode: PropTypes.func.isRequired,
  mode: PropTypes.string.isRequired,
  onEnter: PropTypes.func.isRequired,
  viewer: PropTypes.object.isRequired,
  BIMWINNER: PropTypes.object.isRequired,
};

const mapStateToProps = (state) => ({
  mode: state.button.mode,
  viewer: state.system.viewer3D,
  BIMWINNER: state.system.BIMWINNER,
});
const mapDispatchToProps = (dispatch) => ({
  changeMode: mode => {
    dispatch(changeMode(mode));
  }
});
const WrappedContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(FreeSection);

export default WrappedContainer;
