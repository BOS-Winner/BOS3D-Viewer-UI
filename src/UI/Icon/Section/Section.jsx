import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import Icon from "Base/Icon";
import Modal from "Base/Modal";
import { changeMode } from "../action";
import toastr from "../../toastr";
import style from "./style.less";
import fuckIE from "../../theme/fuckIE.less";
import sectionPng from "../img/white/modelSection.png";
import resetPng from "../img/white/reset2.png";
import hidePng from "../img/white/hide@2x.png";
import showPng from "../img/white/show@2x.png";

class Section extends React.Component {
  constructor(props) {
    super(props);
    this.formRef = React.createRef();
    this.state = {
      dialogVisible: false,
      isShowSectionBox: true
    };
  }

  componentDidUpdate(prevProps) {
    // 开启剖切模式
    if (prevProps.mode !== "剖切模式" && this.props.mode === "剖切模式") {
      if (prevProps.mode !== '') {
        toastr.info(`剖切模式下将关闭${prevProps.mode}`, "", {
          target: `#${this.props.viewer.viewport}`
        });
      }
      this.props.viewer.enableSectionBox();
      this.props.viewer.setSectionBoxMode("normal");
      this.props.viewer.render();
      this.props.onEnter();
    }
    // 关闭剖切模式
    if (prevProps.mode === "剖切模式" && this.props.mode !== "剖切模式") {
      this.props.viewer.disableSectionBox();
      this.props.viewer.render();
    }
  }

  changeDialog(visible) {
    if (!visible) {
      this.props.changeMode('');
    }
    this.setState({
      dialogVisible: visible
    });
  }

  onClick() {
    const selected = (this.props.mode === "剖切模式");
    this.setState(() => {
      this.props.changeMode(selected ? "" : "剖切模式");
      return {
        dialogVisible: !selected,
      };
    });
  }

  toggleSectionBox() {
    const visible = this.state.isShowSectionBox;
    if (visible) {
      this.props.viewer.hideSectionBox();
    } else {
      this.props.viewer.showSectionBox();
    }
    this.setState({
      isShowSectionBox: !visible
    });
  }

  resetSectionBox() {
    this.props.viewer.resetSectionBox();
    this.setState({
      isShowSectionBox: true,
    });
  }

  render() {
    const selected = (this.props.mode === "剖切模式");
    const { isShowSectionBox } = this.state;
    return (
      <div
        className={selected ? fuckIE.select : ''}
        title="模型剖切"
        role="button"
        tabIndex={0}
        onClick={() => { this.onClick() }}
      >
        <Icon img={sectionPng} />
        <Modal
          visible={this.state.dialogVisible && selected}
          onCancel={() => { this.changeDialog(false) }}
          title="模型剖切"
          destroyOnClose
          width="160px"
          height="100px"
          minWidth={155}
          minHeight={90}
          viewportDiv={this.props.viewer.viewportDiv}
        >
          <div className={style.switch}>
            <div
              className={style.reset}
              title={isShowSectionBox ? "隐藏剖切盒" : '显示剖切盒'}
              role="presentation"
              onClick={() => this.toggleSectionBox()}
              style={{
                backgroundImage: `url(${isShowSectionBox ? hidePng : showPng})`
              }}
            />
            <div
              className={style.eye}
              title="重置剖切状态"
              role="presentation"
              onClick={() => this.resetSectionBox()}
              style={{
                backgroundImage: `url(${resetPng})`
              }}
            />
          </div>
        </Modal>
      </div>
    );
  }
}

Section.propTypes = {
  changeMode: PropTypes.func.isRequired,
  mode: PropTypes.string.isRequired,
  onEnter: PropTypes.func.isRequired,
  viewer: PropTypes.object.isRequired,
};

const mapStateToProps = (state) => ({
  mode: state.button.mode,
  viewer: state.system.viewer3D,
});
const mapDispatchToProps = (dispatch) => ({
  changeMode: mode => {
    dispatch(changeMode(mode));
  }
});
const WrappedContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(Section);

export default WrappedContainer;
