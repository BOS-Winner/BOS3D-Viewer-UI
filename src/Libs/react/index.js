import React from "react";
import useUpdateEffect from "./useUpdateEffect";

React.useUpdateEffect = useUpdateEffect;

export default React;
// export { default as useUpdateEffect } from "./useUpdateEffect";
