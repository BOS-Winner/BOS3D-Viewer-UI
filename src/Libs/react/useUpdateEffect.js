import { useEffect, useRef } from "react";

export default function (effect, deps) {
  const isFirstRender = useRef(true);
  useEffect(() => {
    if (isFirstRender.current) {
      isFirstRender.current = false;
    } else {
      effect();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, deps);
}
