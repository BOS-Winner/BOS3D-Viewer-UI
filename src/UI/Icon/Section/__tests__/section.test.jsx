/* eslint-disable react/prop-types */
import React from 'react';
import { Provider } from 'react-redux';
import { genStore } from 'jest/initTest';
import { render, act } from "@testing-library/react";
import Section from '../Section';

jest.mock('Base/Modal', () => props => (
  <div
    className={props.className}
    style={props.style}
  >
    {props.children}
  </div>
));
const onEnter = jest.fn();
const store = genStore();
let section;
let getBtnByTitle;

describe('section', () => {
  beforeEach(() => {
    const { container, getByTitle } = render((
      <Provider store={store}>
        <Section onEnter={onEnter} />
      </Provider>
    ));
    section = container;
    getBtnByTitle = getByTitle;
    const viewer3D = store.getState().system.viewer3D;
    viewer3D.hideSectionBox = jest.fn();
    viewer3D.showSectionBox = jest.fn();
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  it('should switch mode', () => {
    expect(section).toMatchSnapshot();
    act(() => {
      getBtnByTitle('模型剖切').click();
    });
    expect(store.getState().button.mode).toEqual('剖切模式');
    expect(section).toMatchSnapshot();
    expect(onEnter).toBeCalledTimes(1);
    act(() => {
      getBtnByTitle('模型剖切').click();
    });
    expect(store.getState().button.mode).toEqual('');
    expect(section).toMatchSnapshot();
  });

  it("should toggle section box", () => {
    const viewer3D = store.getState().system.viewer3D;
    act(() => {
      getBtnByTitle('隐藏剖切盒').click();
    });
    expect(viewer3D.hideSectionBox).toBeCalledTimes(1);
    expect(section).toMatchSnapshot();
    act(() => {
      getBtnByTitle('显示剖切盒').click();
    });
    expect(viewer3D.showSectionBox).toBeCalledTimes(1);
    expect(section).toMatchSnapshot();
  });

  it('should reset section box', () => {
    const viewer3D = store.getState().system.viewer3D;
    viewer3D.resetSectionBox = jest.fn();
    act(() => {
      getBtnByTitle('重置剖切状态').click();
    });
    expect(viewer3D.resetSectionBox).toBeCalledTimes(1);
  });
});
