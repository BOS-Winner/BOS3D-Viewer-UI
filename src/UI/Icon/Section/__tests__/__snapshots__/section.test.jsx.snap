// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`section should switch mode 1`] = `
<div>
  <div
    class=""
    role="button"
    tabindex="0"
    title="模型剖切"
  >
    <div
      class="icon "
      data-selected="false"
      role="button"
      style="background-image: url(modelSection.png);"
      tabindex="0"
    />
    <div>
      <div
        class="switch"
      >
        <div
          class="reset"
          role="presentation"
          style="background-image: url(hide@2x.png);"
          title="隐藏剖切盒"
        />
        <div
          class="eye"
          role="presentation"
          style="background-image: url(reset2.png);"
          title="重置剖切状态"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`section should switch mode 2`] = `
<div>
  <div
    class="select"
    role="button"
    tabindex="0"
    title="模型剖切"
  >
    <div
      class="icon "
      data-selected="false"
      role="button"
      style="background-image: url(modelSection.png);"
      tabindex="0"
    />
    <div>
      <div
        class="switch"
      >
        <div
          class="reset"
          role="presentation"
          style="background-image: url(hide@2x.png);"
          title="隐藏剖切盒"
        />
        <div
          class="eye"
          role="presentation"
          style="background-image: url(reset2.png);"
          title="重置剖切状态"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`section should switch mode 3`] = `
<div>
  <div
    class=""
    role="button"
    tabindex="0"
    title="模型剖切"
  >
    <div
      class="icon "
      data-selected="false"
      role="button"
      style="background-image: url(modelSection.png);"
      tabindex="0"
    />
    <div>
      <div
        class="switch"
      >
        <div
          class="reset"
          role="presentation"
          style="background-image: url(hide@2x.png);"
          title="隐藏剖切盒"
        />
        <div
          class="eye"
          role="presentation"
          style="background-image: url(reset2.png);"
          title="重置剖切状态"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`section should toggle section box 1`] = `
<div>
  <div
    class="select"
    role="button"
    tabindex="0"
    title="模型剖切"
  >
    <div
      class="icon "
      data-selected="false"
      role="button"
      style="background-image: url(modelSection.png);"
      tabindex="0"
    />
    <div>
      <div
        class="switch"
      >
        <div
          class="reset"
          role="presentation"
          style="background-image: url(show@2x.png);"
          title="显示剖切盒"
        />
        <div
          class="eye"
          role="presentation"
          style="background-image: url(reset2.png);"
          title="重置剖切状态"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`section should toggle section box 2`] = `
<div>
  <div
    class=""
    role="button"
    tabindex="0"
    title="模型剖切"
  >
    <div
      class="icon "
      data-selected="false"
      role="button"
      style="background-image: url(modelSection.png);"
      tabindex="0"
    />
    <div>
      <div
        class="switch"
      >
        <div
          class="reset"
          role="presentation"
          style="background-image: url(hide@2x.png);"
          title="隐藏剖切盒"
        />
        <div
          class="eye"
          role="presentation"
          style="background-image: url(reset2.png);"
          title="重置剖切状态"
        />
      </div>
    </div>
  </div>
</div>
`;
